---
title: "*cellcount*: an open-source and automated approach to cyanobacteria enumeration"
author:
  - name: Tyler E. Harman 
    affiliation: Consolidated Safety Services, Fairfax, VA 22031, USA
      National Oceanic and Atmospheric Administration, National Centers for Coastal Ocean Science, Beaufort, NC 28516, USA
    affiliation_email: tyler.harman@noaa.gov
    orcid_id: 0000-0001-5007-8170
    
  - name: Mark W. Vandersea
    affiliation: National Oceanic and Atmospheric Administration, National Centers for Coastal Ocean Science, Beaufort, NC 28516, USA
    
  - name: David E. Berthold
    affiliation: Department of Biological Sciences, Florida International University, Miami, FL 33199, USA
  
  - name: H. Dail Laughinghouse
    affiliation: Department of Biological Sciences, Florida International University, Miami, FL 33199, USA
    orcid_id: 0000-0003-1018-6948
  
  - name: Brandon Boyd
    affiliation: U.S. Army Corps of Engineers, Engineer Research and Development Center, Vicksburg, MS 39180, USA
  
  - name: Kaytee Pokrzywinski
    affiliation: National Oceanic and Atmospheric Administration, National Centers for Coastal Ocean Science, Beaufort, NC 28516, USA
date: "`r Sys.Date()`"
output: distill::distill_article
---
# Summary

The ***cellcount*** package is a continuation of the report outlined in [@Pokrzywinski2019] and simplfies the quantification of bacteria and cyanobacteria *via* flourescence microscopy, using the computing language [R](https://www.r-project.org/). Functions within this package deconstruct images to assess individual cells, generate cell counts, and assess overall cell density from samples. In addition, additional tools are presented in this package using [R Shiny](https://cran.r-project.org/package=shiny) [@Chang2022] interfaces to demonstrate ease-of-use for researchers not familiar with the R environment. All materials outlined in this article, along with tutorials, are found in [GitHub.](https://github.com/teharman/cellcount_main)

# Statement of need

Harmful algal blooms (HABs) can exist within freshwater, estuarine, and marine environments. These can occur naturally within the 
environment and are stabilized in terms of population by means of nutrient availability, temperature, light, and competition with other
phytoplankton [@Catherine2013]. However, in recent years, HABs have become increasingly common throughout the world, where climate change and anthropogenic impacts exacerbate the commonality of HABs. These blooms can cause harm to the aquatic environment, reducing phytoplankton assemblages, causing anoxic environments, and toxin production impacting higher trophic level organisms [@Kurmayer2003], [@Stumpf2012], [@Catherine2013], [@Borges2019], and [@Wood2020]. From a public health aspect, the production of toxins can cause health issues or even death, but also cause impact from an economic standpoint such as reductions in property value and tourism in coastal areas [@Carey2012].
  
Several methods have been established to understand HAB dynamics, primarily to develop monitoring protocols and establish mitigation 
practices. Microscopic enumeration has been considered as the gold standard for several decades as the method is widely available and simple to use [@Jin2018]. Although it is the standard method, microscopic enumeration can be time consuming, can be limited to small sample quantities, and requires significant knowledge of species identification [@Li2014] and [@Jin2018]. Technology has become significant in advancing scientific methods, and it is no less significant in its role in advancing enumeration processes. Several commercially available products such as the FlowCam (Fluid Imaging Technologies Inc.) and Imaging FlowCytobot (McLane Research Laboratories Inc.) can be used to simplify enumeration methods (REF). These technologies can be accurate, demonstrate time-savings, and simplify data collection (REF). However, several limitations can be observed such as particle size, misidentification of cells, and overall grouping of particles within samples, which can require assistance from researchers (REF). In addition, these technologies are expensive to purchase and maintain, reducing the nature of these technologies being widely available [@Stauffer2019]. A few studies have explored the process of automated cell enumeration methods by means of technology and software development, but promoting widespread availability. ImageJ is an open-source imaging analysis software and has available software for particle/cell enumeration [@Schulze2011]. Other software algorithms have been developed to accomplish automated cellular enumeration [@Jin2018]. These methods can still be considered time-consuming (ImageJ) and require extensive knowledge of mathematical modeling [@Jin2018].

The methods mentioned previously can be viable options for researchers regarding monitoring purposes, but can be expensive and require 
extensive training and materials. This leads to the suggestion of the need for accurate, low-cost, and widely available methods for rapid and simplistic automated enumeration of cells. A report released by [@Pokrzywinski2019] outlines an automated cyanobacteria enumeration script using the open-source programming language, R. [@Pokrzywinski2019] shows that while using fluorescent microscopy images, this enumeration script performs considerably well compared to standard microscopic enumeration methods. However, the previous presentation of this script can be considered extremely difficult to grasp for inexperienced R users. We utilized this script as a template for the development of an R package, involving the reduction of code into simple functions, expanding the ability for end-user modifications, development of a vignette for inexperienced users, and promoting open-source availability.

# Overview of functions

The ***cellcount*** package is an open-source tool that can be accessed by researchers for use in the computing language R. Readers should note that the Bioconductor package ***EBImage*** [@Pau2010], ***tiff*** [@Urbanek2022], ***raster*** [@Hijmans2023], and ***pixmap*** [@Bivand2021] packages are needed for use of this toolset for accurate data collection.

The ***cellcount*** package can be installed locally and used within the [RStudio](https://posit.co/download/rstudio-desktop/) environment. The ***cellcount*** functions are to be used in sequential order, as outlined in the *sandbox_analysis_test.R* script embedded within the ***cellcount*** package. Once image files have been identified, users can begin to upload their images using ***tiff*** package functions (see *readTIFF*). From here, images within the R environment must be converted into formats that R can use to process images. The *greyscale* function converts 24-bit images (RGB) into 8-bit images, and contains a built-in numerical contrast adjustment.

```{r,include=FALSE}
library(EBImage)
library(tiff)
library(pixmap)
library(raster)
library(beepr)
library(knitr)
Cell.Count <- data.frame(Image_File_Name = character(0), Cell_Count = numeric(0))
savdir <- ("C:/Users/Tyler.Harman/Desktop/cellcount_work/cellcount_main/test_data/")
image_savdir <- ("C:/Users/Tyler.Harman/Desktop/cellcount_work/cellcount_main/test_results/")
images <- list.files("C:/Users/Tyler.Harman/Desktop/cellcount_work/cellcount_main/test_images/Microcystis_F192_40x/", pattern = "tif", full.name = T)
images_names <- list.files("C:/Users/Tyler.Harman/Desktop/cellcount_work/cellcount_main/test_images/Microcystis_F192_40x/", pattern = "tif", full.name = F)
file_name<-paste0("F192_")
imgNames <- paste0(file_name, images_names)
read_images <- lapply(images, readTIFF)
img_transposed <- lapply(read_images,aperm,c(2,1,3))
names(images) <- imgNames
```

```{r,message=FALSE}
greyscale <- function(x, contrast = 2) {
  x <- contrast * x
  x <- x[, , 1] + x[, , 2] + x[, , 3]
  x <- x / max(x)
  x <- normalize(x, inputRange = c(0.1, 0.75))
  return(x)
}
```
```{r,echo=FALSE,fig.cap="\\label Converted greyscale image of *Microcystis* sp."}
grey_images <- lapply(img_transposed, greyscale, contrast = 1)
display(grey_images[[5]])
```

The *mapped* function is used to turn 8-bit images into a data matrix from the dimensions of the initial image, using the "*as.matrix*" function (see the ***data.table*** package), and includes a pixel intensity threshold adjustment. This removes any pixel that is under the threshold value to easily separate the background from cells.

```{r, message=FALSE}
mapped <- function(x, threshold = 0.3) {
  x <- as.matrix(x)
  x[x < threshold] <- 0
  return(x)
}
```
```{r,include=FALSE}
imagesMapped <- lapply(grey_images, mapped, threshold = 0.3)
```

The *border* function is built into several other functions within ***cellcount*** (i.e. *countImages*, *image_convert*, etc.), where its primarily goal is to predict cells along a predetermined border and to remove these cells from the final cell count estimate from each image. (WHY IS THAT IMPORTANT-OUTLINE IT HERE).

```{r, message=FALSE}
border <- function(x, y) {
  c(x[1:y[1], 1], x[1:y[1], y[2]], x[1, 1:y[2]], x[y[1], 1:y[2]])
}
```

The *image_convert* functions take the data matrices corresponding to each image and counts individual cells within an image, omitting the cells indentified from the *border* function. Fluorescence microscopy images analyzed by this function are of a single fluorescent dimension (i.e. auto-fluorescence, DAPI, etc.). End-users should note that this function is useful only for analyzing images containing single-cell cyanobacteria. *fillHull* is placed first in this function before *watershed* in order to disseminate single-cells with more accuracy. A parallel function in ***cellcount***, *image_convert2*, is similarly organized but meant for filamentous or colonial-type cyanobacteria. The difference with this is largely where *watershed* is placed before *fillHull*, opposite of *image_convert*, which improves accuracy. Within these two functions, several variables can be adjusted by end-users, such as the width and height of the rectangular window (see ***EBImage***), the pixel area threshold (i.e., removes cells or other objects under a certain pixel area), and the tolerance and ‘ext’ variables (see ***EBImage***).

```{r}
image_convert <- function(x, w = 17, h = 17, offset = 0.001, areathresh = 50, tolerance= 0.5, ext = 1) {
  image <- thresh(x, w = w, h = h, offset = offset)
  image1 <- fillHull(image)
  image2 <- watershed(distmap(image1), tolerance = tolerance, ext = ext)
  nf <- computeFeatures.shape(image2)
  nr <- which(nf[, "s.area"] < areathresh)
  image3 <- rmObjects(image2, nr)
  return(image3)
}
```
```{r}
image_convert2 <- function(x, w = 17, h = 17, offset = 0.001, areathresh = 50, tolerance= 0.5, ext = 1) {
  image <- thresh(x, w = w, h = h, offset = offset)
  image1 <- watershed(distmap(image), tolerance = tolerance, ext = ext)
  image2 <- fillHull(image1)
  nf <- computeFeatures.shape(image2)
  nr <- which(nf[, "s.area"] < areathresh)
  image3 <- rmObjects(image2, nr)
  return(image3)
}
```

Once the conversion occurs, the *countImages* function generates a color-label image, where each separated cell is given a unique color-label. This function includes removing cells along a predetermined border using the *border* function in order to improve count accuracy. End-users may view these color-label images using the *display* function (see the ***arm*** package).

```{r}
countImages <- function(image, normalize = TRUE, removeEdgeCells = TRUE) {
  if (removeEdgeCells) {
    dims <- dim(image)
    border1 <- border(image, dims)
    ids <- unique(border1[which(border1 != 0)])
    inner <- rmObjects(image, ids)
    EBImage::colorLabels(inner, normalize = normalize)
  } else {
    EBImage::colorLabels(image, normalize = normalize)
  }
}
```
```{r,echo=FALSE,fig.cap="A converted color-label image of *Microcystis* sp."}
imagesConverted <- image_convert(imagesMapped[[5]], w = 10, h = 10, offset = 0.001, areathresh = 250, tolerance = 0.8, ext = 1)
final_img <- countImages(imagesConverted, normalize = T, removeEdgeCells = T)
display(final_img)
```

The final function, *countCells*, counts the number of cells resulting from the image generated by the *countImages* function. Once again, the *border* function is integrated here to omit counting cells within its predetermined border. Other items within the *sandbox_analysis_test.R* script enables users to analyze images in bulk, generate tables of results, calculate cell densities using equations from [@Wetzel2000], and export data to .csv files for easy organization. 

```{r}
countCells <- function(x) {
  countAll <- max(x)
  dims <- dim(x)
  border1 <- border(x, dims)
  ids <- unique(border1[which(border1 != 0)])
  countInner <- countAll - length(ids)
  return(countInner)
}
```

```{r,echo=FALSE,fig.pos="h"}
for (z in 1:length(images)) {
  imagesConverted <- image_convert(imagesMapped[[z]], w = 10, h = 10, offset = 0.001, areathresh = 250, tolerance = 0.8, ext = 1)
  final_img <- countImages(imagesConverted, normalize = T, removeEdgeCells = T)
  count <- countCells(imagesConverted)
  Cell.Count[nrow(Cell.Count) + 1, ] <- c(imgNames[[z]], count)
}
cell.total <- as.numeric(Cell.Count$Cell_Count)
cell.total <- sum(cell.total)
Cell.Count[nrow(Cell.Count) + 1, ] <- c(" ", " ")
Cell.Count[nrow(Cell.Count) + 1, ] <- c("Cell Total", cell.total)
cell.av<-(cell.total/10) #change this via total number of images
Cell.Count[nrow(Cell.Count) + 1, ] <- c("Imaging Average",cell.av)
cell_density <- function(x, FOV = 0.0726, images = 10, filtration.area = 213.8, volume = 1, total.volume = 5) {
  MCF <- filtration.area / FOV
  N <- x / images
  cyano <- (MCF * N) / volume
  cyano_total <- cyano * total.volume
  return(cyano_total)
}
cell.den <- cell_density(cell.total, FOV = 0.078125, images = 10, filtration.area = 213.8, volume = 1, total.volume = 5)
Cell.Count[nrow(Cell.Count) + 1, ] <- c("Cell Density", cell.den)
cell.mL<-(cell.den/5) #change this via total volume
Cell.Count[nrow(Cell.Count) + 1, ] <- c("Cells per mL",cell.mL)
kable(Cell.Count, caption = "Analyzed results from test images of *Microcystis* sp.")
```

# Graphical user interface using Shiny apps

Informal discussions with researchers ignited the development of a graphical user interface (GUI) using Shiny apps, paired with functions from the ***cellcount*** package. This results in the use of two functions, *input_data* and *analyze images*, to run the entire analysis process as outlined in the *sandbox_analysis_test.R* script. In addition to packages required by ***cellcount***, these GUIs require the ***shiny*** [@Chang2022], ***shinyalert*** [@1Attali2021], ***shinyFiles*** [@Pedersen2022], ***shinyjs*** [@2Attali2021], and ***beepr*** [@Baath2018] libraries.

*input_data* uses a Shiny app UI for end-users to easily fill out the needed information in order to correctly label analyzed images, provide numerical inputs for variables, select image analysis type, and easily select directories containing images and saving data. The UI format is a simple approach, with tabs outlined on the left-hand side of the interface for its respective purpose (i.e., directory selection, analysis type selection, etc.). Each tab contains interactive buttons or drop-down menus for users to enter in the respective information or make appropriate selections for their analysis needs. Shiny directory buttons provide users a simple and familiar interface to select appropriate folders containing images or to set up save directories. Alert modals (see ***shinyalert***) are used for users to interact and provide information in text format, with helpful dialog to keep users organized and informed of the information needed. Once all information is provided, users select the "Completion" tab and click the "submit data entries" button to complete the data input process.

(INSERT IMAGES OF *input_data* SHINY APP HERE)

*analyze_images* uses a simplistic Shiny UI which consists of two buttons, one to run the analysis and the other to close the UI interface. 

# Limitiations and warnings

# Conclusions

# Acknowledgements

# Literature Cited

---
references:
  - type: article-journal
    id: Catherine2013
    author:
      - family: Catherine
        given: Q.
      - family: Susanna
        given: W.
      - family: Isidora
        given: E.S.
      - family: Mark
        given: H.
      - family: Aurelie
        given: V.
      - family: Jean-Francois
        given: H.
    issued:
      date-parts:
      - - 2013
        - 10
        - 1
    title: 'A review of current knowledge on toxic benthic freshwater cyanobacteria - Ecology, toxin production and risk management'
    container-title: Water Research
    volume: 47
    issue: 15
    page: 5464-5479
    DOI: 10.1016/j.watres.2013.06.042
    URL: https://www.sciencedirect.com/science/article/pii/S0043135413005332
    
  - type: article-journal
    id: Carey2012
    author:
      - family: Carey
        given: C.C.
      - family: Ibelings
        given: B.W.
      - family: Hoffmann
        given: E.P.
      - family: Hamilton
        given: D.P.
      - family: Brookes
        given: J.D.
    issued:
      date-parts:
      - - 2012
        - 4
        - 1
    title: 'Eco-physiological adaptations that favour freshwater cyanobacteria in a changing climate'
    container-title: Water Research
    volume: 46
    issue: 5
    page: 1394-1407
    DOI: 10.1016/j.watres.2011.12.016
    URL: https://www.sciencedirect.com/science/article/pii/S0043135411007883
    
  - type: article-journal
    id: Kurmayer2003
    author:
      - family: Kurmayer
        given: R.
      - family: Christiansen
        given: G.
      - family: Chorus
        given: I.
    issued:
      date-parts:
      - - 2003
        - 2
    title: 'The Abundance of Microcystin-Producing Genotypes Correlates Positively with Colony Size in Microcystis sp. and 
      Determines Its Microcystin Net Production in Lake Wannsee'
    container-title: Applied and Evnrionmental Microbiology
    volume: 69
    issue: 2
    page: 787-795
    DOI: 10.1128/AEM.69.2.787-795.2003
    URL: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC143648/
    
  - type: article-journal
    id: Stumpf2012
    author:
      - family: Stumpf
        given: R.P.
      - family: Wynne
        given: T.T.
      - family: Baker
        given: D.B.
      - family: Fahnenstiel
        given: G.L.
    issued:
      date-parts:
      - - 2012
        - 8
        - 1
    title: 'Interannual Variability of Cyanobacterial Blooms in Lake Erie'
    container-title: PLoS One
    volume: 7
    issue: 8
    page: e42444
    DOI: 10.1371/journal.pone.0042444
    URL: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0042444
    
  - type: article-journal
    id: Borges2019
    author:
      - family: Borges
        given: H.D.
      - family: Cicerelli
        given: R.E.
      - family: de Almeida
        given: T.
      - family: Roig
        given: H.L.
      - family: Olivetti
        given: D.
    issued:
      date-parts:
      - - 2019
        - 5
        - 7
    title: 'Monitoring cyanobacteria occurrence in freshwater reservoirs using semi-analytical algorithms and orbital remote sensing'
    container-title: Marine and Freshwater Research
    volume: 71
    issue: 5
    page: 569-578
    DOI: 10.1071/MF18377
    URL: https://www.publish.csiro.au/MF/MF18377
    
  - type: article-journal
    id: Wood2020
    author:
      - family: Wood
        given: S.A.
      - family: Kelly
        given: L.
      - family: Bouma-Gregson
        given: K.
      - family: Humbert
        given: J.F.
      - family: Laughinghouse
        given: H.D.
      - family: Lazorchak
        given: J.
      - family: McAllister
        given: T.       
      - family: McQueen
        given: A.
      - family: Pokrzywinski
        given: K.
      - family: Puddick
        given: J.
      - family: Quiblier
        given: C.
      - family: Reitz
        given: L.A.
      - family: Ryan
        given: K.
      - family: Vadeboncoeur
        given: Y.
      - family: Zastepa
        given: A.
      - family: Davis
        given: T.W.
    issued:
      date-parts:
      - - 2020
        - 10
        - 1
    title: 'Toxic benthic freshwater cyanobacterial proliferations: Challenges and solutions for enhancing knowledge and improving 
      monitoring and mitigation'
    container-title: Freshwater Biology
    volume: 65
    issue: 10
    page: 1824-1842
    DOI: 10.1111/fwb.13532
    URL: https://onlinelibrary.wiley.com/doi/10.1111/fwb.13532
    
  - type: article-journal
    id: Jin2018
    author:
      - family: Jin
        given: C.
      - family: Mesquita
        given: M.M.F.
      - family: Deglint
        given: J.L.
      - family: Emelko
        given: M.B.
      - family: Wong
        given: A.
    issued:
      date-parts:
      - - 2018
        - 6
        - 13
    title: 'Quantification of cyanobacterial cells via a novel imaging-driven technique with an integrated fluorescence signature'
    container-title: Scientific Reports
    volume: 8
    DOI: 10.1038/s41598-018-27406-0
    URL: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5998128/
    
  - type: article-journal
    id: Li2014
    author:
      - family: Li
        given: M.
      - family: Zhu
        given: W.
      - family: Gao
        given: L.
    issued:
      date-parts:
      - - 2014
        - 2
        - 26
    title: 'Analysis of Cell Concentration, Volume Concentration, and Colony Size of Microcystis Via Laser Particle Analyzer'
    container-title: Environmental Management
    volume: 53
    page: 947-958
    DOI: 10.1007/s00267-014-0252-8
    URL: https://link.springer.com/article/10.1007/s00267-014-0252-8
    
  - type: article-journal
    id: Stauffer2019
    author:
      - family: Stauffer
        given: B.A.
      - family: Bowers
        given: H.A.
      - family: Buckley
        given: E.
      - family: Davis
        given: T.W.
      - family: Johengen
        given: T.H.
      - family: Kudela
        given: R.
      - family: McManus
        given: M.A. 
      - family: Purcell
        given: H.
      - family: Smith
        given: G.J.
      - family: Vander Woude
        given: A.
      - family: Tamburri
        given: M.N.
    issued:
      date-parts:
      - - 2019
        - 7
        - 16
    title: 'Considerations in Harmful Algal Bloom Research and Monitoring: Perspectives From a Consensus-Building Workshop and Technology 
      Testing'
    container-title: Frontiers in Marine Science
    volume: 6
    issue: 399
    DOI: 10.3389/fmars.2019.00399
    URL: https://www.frontiersin.org/articles/10.3389/fmars.2019.00399/full
    
  - type: technical-report
    id: Pokrzywinski2019
    author:
      - family: Pokrzywinski
        given: K.
      - family: Boyd
        given: B.M.
      - family: Smith
        given: S.J.
    issued:
      date-parts:
      - - 2019
        - 8
    title: 'A high-throughput method for counting cyanobacteria using fluorescence microscopy'
    DOI: 10.21079/11681/33870
    URL: https://erdc-library.erdc.dren.mil/jspui/handle/11681/33870
    
  - type: article-journal
    id: Schulze2011
    author:
      - family: Schulze
        given: K.
      - family: Lopez
        given: D.A.
      - family: Tillich
        given: U.M.
      - family: Frohme
        given: M.
    issued:
      date-parts:
      - - 2011
        - 11
        - 30
    title: 'A simple viability analysis for unicellular cyanobacteria using a new autofluorescence assay, automated microscopy, and ImageJ'
    container-title: BMC Biotechnology
    volume: 11
    issue: 118
    DOI: 10.1186/1472-6750-11-118
    URL: https://link.springer.com/article/10.1186/1472-6750-11-118
    
  - type: article-journal
    id: Pau2010
    author:
      - family: Pau
        given: G.
      - family: Fuchs
        given: F.
      - family: Skylar
        given: O.
      - family: Boutros
        given: M.
      - family: Huber
        given: W.
    issued:
      date-parts:
      - - 2010
        - 3
        - 23
    title: 'EBImage—an R package for image processing with applications to cellular phenotypes'
    container-title: Bioinformatics
    volume: 26
    issue: 7
    page: 979-981
    DOI: 10.1093/bioinformatics/btq046
    URL: https://academic.oup.com/bioinformatics/article/26/7/979/211835
    
  - type: package
    id: Urbanek2022
    author:
      - family: Urbanek
        given: S.
      - family: Johnson
        given: K.
    issued:
      date-parts:
      - - 2022
        - 1
        - 31
    title: 'tiff: Read and Write TIFF Images. R package version 0.1-11.'
    URL: https://www.rforge.net/tiff/
    
  - type: package
    id: Bivand2021
    author:
      - family: Bivand
        given: R.
      - family: Leisch
        given: F.
      - family: Maechler
        given: Martin
    issued:
      date-parts:
      - - 2021
        - 1
        - 29
    title: 'Bitmap Images/Pixel Maps, R Package Version 0.4.12.'
    URL: https://cran.r-project.org/package=pixmap
    
  - type: package
    id: Chang2022
    author:
      - family: Chang
        given: W.
      - family: Cheng
        given: J.
      - family: Allaire
        given: J.
      - family: Boutros
        given: M.
      - family: Huber
        given: W.
    issued:
      date-parts:
      - - 2022
        - 12
        - 15
    title: 'Shiny: Web application framework for R'
    URL: https://cran.r-project.org/web/packages/shiny/index.html
    
  - type: book
    id: Wetzel2000
    author:
      - family: Wetzel
        given: R.G.
      - family: Likens
        given: G.E.
    issued:
      date-parts:
      - - 2000
        - 3
        - 3
    title: 'Limnological Analyses - 3rd Edition'
    page: 271-274
    DOI: 10.1007/978-1-4757-3250-4
    URL: https://link.springer.com/book/10.1007/978-1-4757-3250-4
    
  - type: package
    id: Pedersen2022
    author:
      - family: Pedersen
        given: T.L.
      - family: Nijs
        given: V.
      - family: Schaffner
        given: T.
      - family: Nantz
        given: E.
    issued:
      date-parts:
      - - 2022
        - 8
        - 19
    title: 'shinyFiles: A Server-Side File System Viewer for Shiny'
    URL: https://cran.r-project.org/web/packages/shinyFiles/index.html
    
  - type: package
    id: 1Attali2021
    author:
      - family: Attali
        given: D.
      - family: Edwards
        given: T.
      - family: Wang
        given: Z.
    issued:
      date-parts:
      - - 2021
        - 12
        - 20
    title: 'shinyalert: Easily Create Pretty Popup Messages (Modals) in Shiny'
    URL: https://cran.r-project.org/web/packages/shinyalert/index.html
    
  - type: package
    id: 2Attali2021
    author:
      - family: Attali
        given: D.
    issued:
      date-parts:
      - - 2021
        - 12
        - 23
    title: 'shinyjs: Easily Improve the User Experience of Your Shiny Apps in Seconds'
    URL: https://cran.r-project.org/web/packages/shinyjs/index.html
    
  - type: package
    id: Baath2018
    author:
      - family: Baath
        given: R.
      - family: Dobbyn
        given: A.
    issued:
      date-parts:
      - - 2018
        - 6
        - 4
    title: 'beepr: Easily Play Notification Sounds on any Platform'
    URL: https://cran.r-project.org/web/packages/beepr/index.html
    
  - type: package
    id: Hijmans2023
    author:
      - family: Hijmans
        given: R.J
      - family: van Etten
        given: J.
      - family: Sumner
        given: M.
      - family: Cheng
        given: J.
      - family: Baston
        given: D.
      - family: Bevan
        given: A.
      - family: Bivand
        given: R.
      - family: Busetto
        given: L.
      - family: Canty
        given: M.
      - family: Fasoli
        given: B.
      - family: Forrest
        given: D.
      - family: Ghosh
        given: A.
      - family: Golicher
        given: D.
      - family: Gray
        given: J.
      - family: Greenburg
        given: J.A.
      - family: Hiemstra
        given: P.
      - family: Hingee
        given: K.
      - family: Ilich
        given: A.
      - family: Karney
        given: C.
      - family: Mattiuzzi
        given: M.
      - family: Mosher
        given: S.
      - family: Naimi
        given: B.
      - family: Nowosad
        given: J.
      - family: Pebesma
        given: E.
      - family: Lamigueiro
        given: O.P.
      - family: Racine
        given: E.B.
      - family: Rowlingson
        given: B.
      - family: Shortridge
        given: A.
      - family: Venables
        given: B.
      - family: Wueest
        given: R.
    issued:
      date-parts:
      - - 2023
        - 1
        - 16
    title: 'raster: Geographic Data Analysis and Modeling'
    URL: https://cran.r-project.org/web/packages/raster/index.html
...